### PR TITLE
add widget placeholders feature

### PR DESCRIPTION
**Pull request by ETLaurent**
## Summary

Add placeholder feature.
See [commits](https://github.com/apostrophecms/apostrophe/pull/3910/commits)' PRs for more details.

## What are the specific steps to test this change?

- Checkout `pro-3147-widget-placeholders-e2e` branch on testbed

### Random widget

In `modules/topic-widget/views/widget.html`, change:

```diff
-<h3>My Topic Widget</h3>
+<h3>{{ data.widget.title }}</h3>
```

- Add the `placeholder: true` option to any non-contextual widget (topic for instance)
- Add a `placeholder` attribute to its fields:

```js
title: {
  type: 'string',
  placeholder: 'Topics Placeholder'
}
```

- Add the widget on the homepage area.

Widget should have been inserted with the placeholder (initial modal should not have appeared):

image

Widget wrapper should contain the default `apos-placeholder` class.
Widget should not be rendered on  "Preview".

- Edit the widget:

image

Widget wrapper should not contain the default `apos-placeholder` class.
Widget should be rendered on  "Preview":

image

- Add the `initialModal: false` option to any non-contextual widget (topic for instance)
- Add a `def` attribute to its fields:

```js
      title: {
        type: 'string',
        def: 'Topics def'
      }
```

- Add the widget on the homepage area.

Widget should have been inserted with the def attribute (initial modal should not have appeared):

image

Widget wrapper should not contain the default `placeholder-class-override` class.
Widget should be rendered on  "Preview".

- Add the `placeholder: true`  and `placeholderClass: 'some-class'` options to any non-contextual widget (topic for instance)
- Add a `placeholder` attribute to its fields:

```js
title: {
  type: 'string',
  placeholder: 'Topics Placeholder'
}
``` 

- Add the widget on the homepage area.

Widget should have been inserted with the placeholder (initial modal should not have appeared):

image

Widget wrapper should contain the custom `placeholder-class-override` class.
Widget should not be rendered on  "Preview".

Do the same test for image, video and rich text widgets.
Placeholder for these must be enabled by default.
Adding them to an area should display the placeholder (not the initial edit modal), until edited.
United edited, placeholders should not be rendered on preview.
The image and video widgets placeholder can be overridden with the `placeholderUrl` option, check that it works.
The rich text widget placeholder can be overridden with the `placeholderText` option, check that it works.
Finally, check that disabling the `placeholder` option for these widgets, will display the initial edit modal again when adding them to an area.

## What kind of change does this PR introduce?
*(Check at least one)*

- [ ] Bug fix
- [x] New feature
- [ ] Refactor
- [ ] Documentation
- [ ] Build-related changes
- [ ] Other

## Make sure the PR fulfills these requirements:

- [x] It includes a) the existing issue ID being resolved, b) a convincing reason for adding this feature, or c) a clear description of the bug it resolves
- [x] The changelog is updated
- [ ] Related documentation has been updated
- [x] Related tests have been updated

If adding a new feature without an already open issue, it's best to open a **feature request issue** first and wait for approval before working on it.

**Other information:**

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,16 @@
 # Changelog
 
+## 3.31.0 (UNRELEASED)
+
+### Adds
+
+* Adds the widgets placeholder feature.  
+Non-contextual widgets now take the `placeholder` option. Those having this option set to `true` that are added to an area will be inserted directly on the page with the `placeholder` attribute from their fields as rendered data, without having the initial edit modal displayed.  
+A default `.apos-placeholder` class is added to their wrapper, showing that they are rendered with their placeholder data. This class can be overridden with the `placeholderClass` widget option, in order to apply a custom style on a particular project.
```
Comment (boutell) [inline CHANGELOG.md:9]: The `.` is part of a selector, it is not part of the class name, so leave out the `.` here.

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,5 +1,16 @@
...
 
+## 3.31.0 (UNRELEASED)
+
+### Adds
+
+* Adds the widgets placeholder feature.  
+Non-contextual widgets now take the `placeholder` option. Those having this option set to `true` that are added to an area will be inserted directly on the page with the `placeholder` attribute from their fields as rendered data, without having the initial edit modal displayed.  
+A default `.apos-placeholder` class is added to their wrapper, showing that they are rendered with their placeholder data. This class can be overridden with the `placeholderClass` widget option, in order to apply a custom style on a particular project.  
+Image, Video and Rich Text widgets have the `placeholder` option enabled by default, which can be disabled on project side. Their default placeholder can be overridden thanks to the `placeholderUrl` option (for Image & Video widgets) and `placeholderText` (for Rich Text widget).  
+Placeholders are _never saved_ into the database, and are only rendered on edit mode. When a widget with placeholder(s) is edited, the actual data is saved and replaces the placeholder(s).
```
Comment (boutell) [inline CHANGELOG.md:11]: This is too much information. We should link to the new documentation for this feature, which must be ready to go at the same time as this feature is released.

**Review by boutell — APPROVED** [review]
Looks like a nice clean implementation, thanks for working through all the ins and outs with me.

You raised a question for Stuart here but he's out til November. Maybe Alex can help answer it.

File: CHANGELOG.md (hunk 1 above)
Comment (ETLaurent) [reply CHANGELOG.md:9]: Yeah that's what I thought

**Review by boutell — APPROVED** [review]

**Pull request merged**